packages/std/src/math/signed_decimal.rs

```diff
@@ -0,0 +1,3056 @@
+        assert_eq!(
+            SignedDecimal::from_str("1.5").unwrap(),
+            SignedDecimal::percent(150)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("0.5").unwrap(),
+            SignedDecimal::percent(50)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("0.123").unwrap(),
+            SignedDecimal::permille(123)
+        );
+
+        assert_eq!(
+            SignedDecimal::from_str("40.00").unwrap(),
+            SignedDecimal::percent(4000)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("04.00").unwrap(),
+            SignedDecimal::percent(400)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("00.40").unwrap(),
+            SignedDecimal::percent(40)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("00.04").unwrap(),
+            SignedDecimal::percent(4)
+        );
+        // Negative decimals
+        assert_eq!(
+            SignedDecimal::from_str("-00.04").unwrap(),
+            SignedDecimal::percent(-4)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("-00.40").unwrap(),
+            SignedDecimal::percent(-40)
+        );
+        assert_eq!(
+            SignedDecimal::from_str("-04.00").unwrap(),
+            SignedDecimal::percent(-400)
+        );
+
+        // Can handle DECIMAL_PLACES fractional digits
+        assert_eq!(
+            SignedDecimal::from_str("7.123456789012345678").unwrap(),
+            SignedDecimal(Int128::from(7123456789012345678i128))
+        );
+        assert_eq!(
+            SignedDecimal::from_str("7.999999999999999999").unwrap(),
+            SignedDecimal(Int128::from(7999999999999999999i128))
+        );
+
+        // Works for documented max value
+        assert_eq!(
+            SignedDecimal::from_str("170141183460469231731.687303715884105727").unwrap(),
+            SignedDecimal::MAX
+        );
+        // Works for documented min value
+        assert_eq!(
+            SignedDecimal::from_str("-170141183460469231731.687303715884105728").unwrap(),
+            SignedDecimal::MIN
+        );
+        assert_eq!(
+            SignedDecimal::from_str("-1").unwrap(),
+            SignedDecimal::negative_one()
+        );
+    }
+
+    #[test]
+    fn signed_decimal_from_str_errors_for_broken_whole_part() {
+        match SignedDecimal::from_str("").unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(msg, "Error parsing whole"),
+            e => panic!("Unexpected error: {e:?}"),
+        }
+
+        match SignedDecimal::from_str(" ").unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(msg, "Error parsing whole"),
+            e => panic!("Unexpected error: {e:?}"),
+        }
+
+        match SignedDecimal::from_str("-").unwrap_err() {
+            StdError::GenericErr { msg, .. } => assert_eq!(msg, "Error parsing whole"),
+            e => panic!("Unexpected error: {e:?}"),
+        }
+    }
```

Comment:
I would suggest:
```Rust
    #[test]
    fn signed_decimal_from_str_errors_for_broken_whole_part() {
        let expected_err = StdError::generic_err("Error parsing whole");
        assert_eq!(expected_err, SignedDecimal::from_str("").unwrap_err());
        assert_eq!(expected_err, SignedDecimal::from_str(" ").unwrap_err());
        assert_eq!(expected_err, SignedDecimal::from_str("-").unwrap_err());
    }
```

Reply:
Great stuff, thank you! I'll merge it into this PR.

Reply:
Just fyi: In this repo, we usually put the expected value as the second argument of `assert_eq`.
I cherry picked your tests and changed it already, so no need for you to do it.